File: backend2/src/services/redis/cacheService.ts
/**
 * Cache service for Node.js backend (backend2)
 * Provides high-level caching operations for chat, LLM, embeddings, and jobs
 */

import { RedisClientType } from 'redis';
import { getRedisClient } from './redisClient';
import { logger } from '../utils/logger';
import { maskParsedData } from '../utils/masks';

// TTLs from environment (in seconds)
const TTL_CHAT_CONTEXT = parseInt(process.env.REDIS_TTL_CHAT_CONTEXT || '86400', 10); // 24 hours
const TTL_LLM_CACHE = parseInt(process.env.REDIS_TTL_LLM_CACHE || '604800', 10); // 7 days
const TTL_EMBED_CACHE = parseInt(process.env.REDIS_TTL_EMBED_CACHE || '2592000', 10); // 30 days
const TTL_JOB_STATUS = parseInt(process.env.REDIS_TTL_JOB_STATUS || '172800', 10); // 48 hours
const TTL_SESSION = parseInt(process.env.REDIS_TTL_SESSION || '86400', 10); // 24 hours

// Key prefixes
const PREFIX_CHAT = process.env.REDIS_KEY_PREFIX_CHAT || 'chat:';
const PREFIX_LLM = process.env.REDIS_KEY_PREFIX_LLM || 'llm:cache:';
const PREFIX_EMBED = process.env.REDIS_KEY_PREFIX_EMBED || 'embed:cache:';
const PREFIX_JOB = process.env.REDIS_KEY_PREFIX_JOB || 'job:';
const PREFIX_LOCK = process.env.REDIS_KEY_PREFIX_LOCK || 'locks:';
const PREFIX_FEATURE = process.env.REDIS_KEY_PREFIX_FEATURE || 'feature:session:';

// Chat context window size
const CHAT_CONTEXT_SIZE = parseInt(process.env.REDIS_CHAT_CONTEXT_SIZE || '20', 10);

interface ChatMessage {
  role: string;
  text: string;
  timestamp: string;
}

interface LLMCacheValue {
  reply: string;
  used_docs: string[];
  created_at: string;
}

interface EmbeddingCacheValue {
  embedding: number[];
  text_hash: string;
  created_at: string;
}

interface JobStatus {
  status: 'queued' | 'processing' | 'done' | 'failed';
  progress: number;
  result_ptr?: string;
  updated_at: string;
}

/**
 * High-level cache service
 */
export class CacheService {
  private async getClient(): Promise<RedisClientType | null> {
    return await getRedisClient();
  }

  // Chat Context Operations
  async addChatMessage(chatId: string, role: string, text: string): Promise<boolean> {
    const client = await this.getClient();
    if (!client) return false;

    try {
      const key = `${PREFIX_CHAT}${chatId}:context`;
      const message: ChatMessage = {
        role,
        text,
        timestamp: new Date().toISOString(),
      };

      // Push to left (newest first) and trim to keep last N messages
      await client.lPush(key, JSON.stringify(message));
      await client.lTrim(key, 0, CHAT_CONTEXT_SIZE - 1);
      await client.expire(key, TTL_CHAT_CONTEXT);

      return true;
    } catch (error: any) {
      logger.error({ error, chatId }, 'Failed to add chat message to cache');
      return false;
    }
  }

  async getChatContext(chatId: string, limit?: number): Promise<ChatMessage[]> {
    const client = await this.getClient();
    if (!client) return [];

    try {
      const key = `${PREFIX_CHAT}${chatId}:context`;
      const limitCount = limit || CHAT_CONTEXT_SIZE;

      // Get all messages (or last N)
      const messagesJson = await client.lRange(key, 0, limitCount - 1);
      const messages: ChatMessage[] = [];

      // Reverse to get chronological order (oldest first)
      for (let i = messagesJson.length - 1; i >= 0; i--) {
        try {
          messages.push(JSON.parse(messagesJson[i]));
        } catch (e) {
          logger.warn({ error: e }, 'Failed to parse chat message from cache');
        }
      }

      return messages;
    } catch (error: any) {
      logger.error({ error, chatId }, 'Failed to get chat context from cache');
      return [];
    }
  }

  async setChatMeta(chatId: string, meta: Record<string, any>): Promise<boolean> {
    const client = await this.getClient();
    if (!client) return false;

    try {
      const key = `${PREFIX_CHAT}${chatId}:meta`;
      await client.setEx(key, TTL_CHAT_CONTEXT, JSON.stringify(meta));
      return true;
    } catch (error: any) {
      logger.error({ error, chatId }, 'Failed to set chat meta in cache');
      return false;
    }
  }

  async getChatMeta(chatId: string): Promise<Record<string, any> | null> {
    const client = await this.getClient();
    if (!client) return null;

    try {
      const key = `${PREFIX_CHAT}${chatId}:meta`;
      const value = await client.get(key);
      if (!value) return null;
      return JSON.parse(value);
    } catch (error: any) {
      logger.error({ error, chatId }, 'Failed to get chat meta from cache');
      return null;
    }
  }

  // LLM Cache Operations
  async getLLMCache(modelVersion: string, prompt: string): Promise<LLMCacheValue | null> {
    const client = await this.getClient();
    if (!client) return null;

    try {
      const crypto = await import('crypto');
      const promptHash = crypto.createHash('sha256').update(`${modelVersion}:${prompt}`).digest('hex').slice(0, 16);
      const key = `${PREFIX_LLM}${modelVersion}:${promptHash}`;

      const value = await client.get(key);
      if (!value) return null;

      return JSON.parse(value) as LLMCacheValue;
    } catch (error: any) {
      logger.error({ error }, 'Failed to get LLM cache');
      return null;
    }
  }

  async setLLMCache(
    modelVersion: string,
    prompt: string,
    reply: string,
    usedDocs?: string[]
  ): Promise<boolean> {
    const client = await this.getClient();
    if (!client) return false;

    try {
      const crypto = await import('crypto');
      const promptHash = crypto.createHash('sha256').update(`${modelVersion}:${prompt}`).digest('hex').slice(0, 16);
      const key = `${PREFIX_LLM}${modelVersion}:${promptHash}`;

      const cacheValue: LLMCacheValue = {
        reply,
        used_docs: usedDocs || [],
        created_at: new Date().toISOString(),
      };

      await client.setEx(key, TTL_LLM_CACHE, JSON.stringify(cacheValue));
      return true;
    } catch (error: any) {
      logger.error({ error }, 'Failed to set LLM cache');
      return false;
    }
  }

  // Embedding Cache Operations
  async getEmbeddingCache(text: string): Promise<number[] | null> {
    const client = await this.getClient();
    if (!client) return null;

    try {
      const crypto = await import('crypto');
      const textHash = crypto.createHash('sha256').update(text).digest('hex');
      const key = `${PREFIX_EMBED}${textHash}`;

      const value = await client.get(key);
      if (!value) return null;

      const cached: EmbeddingCacheValue = JSON.parse(value);
      return cached.embedding;
    } catch (error: any) {
      logger.error({ error }, 'Failed to get embedding cache');
      return null;
    }
  }

  async setEmbeddingCache(text: string, embedding: number[]): Promise<boolean> {
    const client = await this.getClient();
    if (!client) return false;

    try {
      const crypto = await import('crypto');
      const textHash = crypto.createHash('sha256').update(text).digest('hex');
      const key = `${PREFIX_EMBED}${textHash}`;

      const cacheValue: EmbeddingCacheValue = {
        embedding,
        text_hash: textHash,
        created_at: new Date().toISOString(),
      };

      await client.setEx(key, TTL_EMBED_CACHE, JSON.stringify(cacheValue));
      return true;
    } catch (error: any) {
      logger.error({ error }, 'Failed to set embedding cache');
      return false;
    }
  }

  // Job Status Operations
  async setJobStatus(
    jobId: string,
    status: 'queued' | 'processing' | 'done' | 'failed',
    progress: number = 0,
    resultPtr?: string
  ): Promise<boolean> {
    const client = await this.getClient();
    if (!client) return false;

    try {
      const key = `${PREFIX_JOB}${jobId}`;

      const jobData: JobStatus = {
        status,
        progress,
        result_ptr: resultPtr,
        updated_at: new Date().toISOString(),
      };

      await client.setEx(key, TTL_JOB_STATUS, JSON.stringify(jobData));
      return true;
    } catch (error: any) {
      logger.error({ error, jobId }, 'Failed to set job status in cache');
      return false;
    }
  }

  async getJobStatus(jobId: string): Promise<JobStatus | null> {
    const client = await this.getClient();
    if (!client) return null;

    try {
      const key = `${PREFIX_JOB}${jobId}`;
      const value = await client.get(key);
      if (!value) return null;

      return JSON.parse(value) as JobStatus;
    } catch (error: any) {
      logger.error({ error, jobId }, 'Failed to get job status from cache');
      return null;
    }
  }

  async deleteJobStatus(jobId: string): Promise<boolean> {
    const client = await this.getClient();
    if (!client) return false;

    try {
      const key = `${PREFIX_JOB}${jobId}`;
      await client.del(key);
      return true;
    } catch (error: any) {
      logger.error({ error, jobId }, 'Failed to delete job status from cache');
      return false;
    }
  }

  // Lock Operations
  async acquireLock(resource: string, timeout: number = 30): Promise<boolean> {
    const client = await this.getClient();
    if (!client) return false;

    try {
      const key = `${PREFIX_LOCK}${resource}`;
      // SET key value NX PX timeout_ms
      const result = await client.set(key, 'locked', {
        NX: true,
        PX: timeout * 1000,
      });
      return result === 'OK';
    } catch (error: any) {
      logger.error({ error, resource }, 'Failed to acquire lock');
      return false;
    }
  }

  async releaseLock(resource: string): Promise<boolean> {
    const client = await this.getClient();
    if (!client) return false;

    try {
      const key = `${PREFIX_LOCK}${resource}`;
      await client.del(key);
      return true;
    } catch (error: any) {
      logger.error({ error, resource }, 'Failed to release lock');
      return false;
    }
  }

  // Feature Flags / Session Preferences
  async setSessionFeatures(sessionId: string, features: Record<string, any>): Promise<boolean> {
    const client = await this.getClient();
    if (!client) return false;

    try {
      const key = `${PREFIX_FEATURE}${sessionId}`;
      // Mask PII before storing
      const maskedFeatures = maskParsedData(features);
      await client.setEx(key, TTL_SESSION, JSON.stringify(maskedFeatures));
      return true;
    } catch (error: any) {
      logger.error({ error, sessionId }, 'Failed to set session features in cache');
      return false;
    }
  }

  async getSessionFeatures(sessionId: string): Promise<Record<string, any> | null> {
    const client = await this.getClient();
    if (!client) return null;

    try {
      const key = `${PREFIX_FEATURE}${sessionId}`;
      const value = await client.get(key);
      if (!value) return null;

      return JSON.parse(value);
    } catch (error: any) {
      logger.error({ error, sessionId }, 'Failed to get session features from cache');
      return null;
    }
  }
}

// Global cache service instance
let cacheService: CacheService | null = null;

export function getCacheService(): CacheService {
  if (!cacheService) {
    cacheService = new CacheService();
  }
  return cacheService;
}

File: backend/app/services/redis/cache_service.py
"""
Cache service for Python backend.
Provides high-level caching operations for chat, LLM, embeddings, and jobs.
"""

import os
import hashlib
import json
import logging
from typing import Optional, List, Dict, Any
from datetime import datetime
from .redis_client import get_redis_client

logger = logging.getLogger(__name__)

# TTLs from environment (in seconds)
TTL_CHAT_CONTEXT = int(os.getenv("REDIS_TTL_CHAT_CONTEXT", "86400"))  # 24 hours
TTL_LLM_CACHE = int(os.getenv("REDIS_TTL_LLM_CACHE", "604800"))  # 7 days
TTL_EMBED_CACHE = int(os.getenv("REDIS_TTL_EMBED_CACHE", "2592000"))  # 30 days
TTL_JOB_STATUS = int(os.getenv("REDIS_TTL_JOB_STATUS", "172800"))  # 48 hours
TTL_SESSION = int(os.getenv("REDIS_TTL_SESSION", "86400"))  # 24 hours

# Key prefixes
PREFIX_CHAT = os.getenv("REDIS_KEY_PREFIX_CHAT", "chat:")
PREFIX_LLM = os.getenv("REDIS_KEY_PREFIX_LLM", "llm:cache:")
PREFIX_EMBED = os.getenv("REDIS_KEY_PREFIX_EMBED", "embed:cache:")
PREFIX_JOB = os.getenv("REDIS_KEY_PREFIX_JOB", "job:")
PREFIX_LOCK = os.getenv("REDIS_KEY_PREFIX_LOCK", "locks:")
PREFIX_FEATURE = os.getenv("REDIS_KEY_PREFIX_FEATURE", "feature:session:")

# Chat context window size
CHAT_CONTEXT_SIZE = int(os.getenv("REDIS_CHAT_CONTEXT_SIZE", "20"))


class CacheService:
    """High-level cache service for common operations."""
    
    def __init__(self):
        self.redis = get_redis_client()
    
    # Chat Context Operations
    def add_chat_message(self, chat_id: str, role: str, text: str) -> bool:
        """
        Add message to chat context (sliding window).
        Maintains last N messages in Redis List.
        """
        key = f"{PREFIX_CHAT}{chat_id}:context"
        message = json.dumps({"role": role, "text": text, "timestamp": datetime.utcnow().isoformat()})
        
        # Push to left (newest first) and trim to keep last N messages
        self.redis.lpush(key, message)
        self.redis.ltrim(key, 0, CHAT_CONTEXT_SIZE - 1)
        
        # Set TTL on the key
        self.redis.get_client().expire(key, TTL_CHAT_CONTEXT)
        
        return True
    
    def get_chat_context(self, chat_id: str, limit: Optional[int] = None) -> List[Dict[str, Any]]:
        """
        Get chat context (last N messages).
        Returns messages in chronological order (oldest first).
        """
        key = f"{PREFIX_CHAT}{chat_id}:context"
        limit = limit or CHAT_CONTEXT_SIZE
        
        # Get all messages (or last N)
        messages_json = self.redis.lrange(key, 0, limit - 1)
        messages = []
        
        for msg_json in reversed(messages_json):  # Reverse to get chronological order
            try:
                messages.append(json.loads(msg_json))
            except json.JSONDecodeError:
                continue
        
        return messages
    
    def set_chat_meta(self, chat_id: str, meta: Dict[str, Any]) -> bool:
        """Set chat metadata (last_active, model_version, etc.)."""
        key = f"{PREFIX_CHAT}{chat_id}:meta"
        return self.redis.set_json(key, meta, ttl=TTL_CHAT_CONTEXT)
    
    def get_chat_meta(self, chat_id: str) -> Optional[Dict[str, Any]]:
        """Get chat metadata."""
        key = f"{PREFIX_CHAT}{chat_id}:meta"
        return self.redis.get_json(key)
    
    # LLM Cache Operations
    def get_llm_cache(self, model_version: str, prompt: str) -> Optional[Dict[str, Any]]:
        """
        Get cached LLM response.
        Key is based on model_version and prompt hash.
        """
        prompt_hash = hashlib.sha256(f"{model_version}:{prompt}".encode()).hexdigest()[:16]
        key = f"{PREFIX_LLM}{model_version}:{prompt_hash}"
        return self.redis.get_json(key)
    
    def set_llm_cache(self, model_version: str, prompt: str, reply: str, used_docs: Optional[List[str]] = None) -> bool:
        """Cache LLM response."""
        prompt_hash = hashlib.sha256(f"{model_version}:{prompt}".encode()).hexdigest()[:16]
        key = f"{PREFIX_LLM}{model_version}:{prompt_hash}"
        
        cache_value = {
            "reply": reply,
            "used_docs": used_docs or [],
            "created_at": datetime.utcnow().isoformat()
        }
        
        return self.redis.set_json(key, cache_value, ttl=TTL_LLM_CACHE)
    
    # Embedding Cache Operations
    def get_embedding_cache(self, text: str) -> Optional[List[float]]:
        """Get cached embedding for text."""
        text_hash = hashlib.sha256(text.encode()).hexdigest()
        key = f"{PREFIX_EMBED}{text_hash}"
        
        cached = self.redis.get_json(key)
        if cached:
            return cached.get("embedding")
        return None
    
    def set_embedding_cache(self, text: str, embedding: List[float]) -> bool:
        """Cache embedding for text."""
        text_hash = hashlib.sha256(text.encode()).hexdigest()
        key = f"{PREFIX_EMBED}{text_hash}"
        
        cache_value = {
            "embedding": embedding,
            "text_hash": text_hash,
            "created_at": datetime.utcnow().isoformat()
        }
        
        return self.redis.set_json(key, cache_value, ttl=TTL_EMBED_CACHE)
    
    # Job Status Operations
    def set_job_status(self, job_id: str, status: str, progress: int = 0, result_ptr: Optional[str] = None) -> bool:
        """
        Set job status in Redis.
        Status: queued, processing, done, failed
        Progress: 0-100
        result_ptr: path to result file or DB id
        """
        key = f"{PREFIX_JOB}{job_id}"
        
        job_data = {
            "status": status,
            "progress": progress,
            "result_ptr": result_ptr,
            "updated_at": datetime.utcnow().isoformat()
        }
        
        return self.redis.set_json(key, job_data, ttl=TTL_JOB_STATUS)
    
    def get_job_status(self, job_id: str) -> Optional[Dict[str, Any]]:
        """Get job status."""
        key = f"{PREFIX_JOB}{job_id}"
        return self.redis.get_json(key)
    
    def delete_job_status(self, job_id: str) -> bool:
        """Delete job status after retrieval."""
        key = f"{PREFIX_JOB}{job_id}"
        return self.redis.delete(key)
    
    # Lock Operations
    def acquire_lock(self, resource: str, timeout: int = 30) -> bool:
        """Acquire distributed lock for resource."""
        key = f"{PREFIX_LOCK}{resource}"
        return self.redis.acquire_lock(key, timeout)
    
    def release_lock(self, resource: str) -> bool:
        """Release distributed lock."""
        key = f"{PREFIX_LOCK}{resource}"
        return self.redis.release_lock(key)
    
    # Feature Flags / Session Preferences
    def set_session_features(self, session_id: str, features: Dict[str, Any]) -> bool:
        """Set user feature preferences for session."""
        key = f"{PREFIX_FEATURE}{session_id}"
        return self.redis.set_json(key, features, ttl=TTL_SESSION)
    
    def get_session_features(self, session_id: str) -> Optional[Dict[str, Any]]:
        """Get user feature preferences."""
        key = f"{PREFIX_FEATURE}{session_id}"
        return self.redis.get_json(key)


# Global cache service instance
_cache_service: Optional[CacheService] = None


def get_cache_service() -> CacheService:
    """Get or create global cache service instance."""
    global _cache_service
    if _cache_service is None:
        _cache_service = CacheService()
    return _cache_service

